File: Data Structures & Algorithms/permutation-string/submission-7.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        c1, c2 = [0] * 26, [0] * 26
        for i in range(len(s1)):
            i1 = ord(s1[i]) - ord('a')
            i2 = ord(s2[i]) - ord('a')
            c1[i1] += 1
            c2[i2] += 1

        if c1 == c2:
            return True

        l = 0
        for r in range(len(s1), len(s2)):
            left = ord(s2[l]) - ord('a')
            right = ord(s2[r]) - ord('a')

            c2[left] -= 1
            c2[right] += 1

            if c1 == c2:
                return True

            l += 1
        
        return False
        

File: Data Structures & Algorithms/three-integer-sum/submission-7.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for(int i = 0; i < nums.length; i++){
            if(i > 0 && (nums[i - 1] == nums[i]))
                continue;
            int left = i + 1, right = nums.length - 1;
            while(left < right){
                if(nums[i] + nums[left] + nums[right] == 0){
                    res.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while(left < right && nums[left] == nums[left + 1])
                        left++;
                    left++;
                    while(left < right && nums[right] == nums[right - 1])
                        right--;
                    right--;
                }

                else if(nums[i] + nums[left] + nums[right] < 0){
                    while(left < right && nums[left] == nums[left + 1])
                        left++;
                    left++;
                }

                else{
                    while(left < right && nums[right] == nums[right - 1])
                        right--;
                    right--;
                }

            }
            
        }
        return res;
    }
}
